clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java

```diff
@@ -95,21 +93,33 @@ public short version() {
         return version;
     }
 
-    public Send toSend(String destination, RequestHeader header) {
-        return new NetworkSend(destination, serialize(header));
+    public ApiKeys apiKey() {
+        return apiKey;
     }
 
-    /**
-     * Use with care, typically {@link #toSend(String, RequestHeader)} should be used instead.
-     */
-    public ByteBuffer serialize(RequestHeader header) {
-        return RequestUtils.serialize(header.toStruct(), toStruct());
+    public final Send toSend(String destination, RequestHeader header) {
+        return SendBuilder.buildRequestSend(destination, header, data());
+    }
+
+    // Visible for testing
+    public final ByteBuffer serializeWithHeader(RequestHeader header) {
```

Comment:
BTW, this method is duplicate to TestUtils#serializeRequestHeader

Reply:
This method serializes the request with the header while the test method serializes the header only. So, not a duplicate, right? Personally, I think it's useful for the class to be symmetric: it can parse from a buffer and serialize to a buffer. Do you have a concern that it may be misused?

Reply:
There is a reasonable argument that `serializeBody` is the counterpart to `parse` though and this particular method doesn't have to be here. I can buy that.

Reply:
> Do you have a concern that it may be misused?
yes. Also the various serialization methods are a bit chaos to me :(

Reply:
Maybe we should move all testing-only serialization to a single utils (in test scope). I grep "static ByteBuffer" + "request/response" and ```MessageUtils```, ```MessageTestUtil```, ```RequestUtils``` and ```TestUtils``` have similar method interface.

Reply:
I have a refactoring that consolidates the logic around `Send`. But I think that would be best done in a separate PR since this PR is really massive. Since these methods are all in internal classes and are only used by tests, I suggest we keep them as they are for now and tackle these nits in the follow up PR.

Reply:
I have these changes locally, so I can submit them immediately after.